lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,817 @@
+  }
+
+  // Checks if a constant of a given size has been declared
+  // If so, its lid will be returned
+  // Otherwise -1 will be returned
+  size_t getConstLID(int64_t val, size_t w) {
+    if (auto it = constToLIDMap.find({val, w}); it != constToLIDMap.end())
+      return it->second;
+
+    // if no lid was found return -1
+    return noLID;
+  }
+
+  /// String generation helper functions
+
+  // Generates a sort declaration instruction given a type (bitvecStr or array)
+  // and a width
+  void genSort(StringRef type, size_t width) {
+    // Check that the sort wasn't already declared
+    if (getSortLID(width) != noLID) {
+      return; // If it has already been declared then return an empty string
+    }
+
+    // Add the width to the declared sorts along with the associated line id
+    sortToLIDMap[width] = lid;
+
+    // Build and return a sort declaration
+    os << lid++ << wsStr << sortStr << wsStr << type << wsStr << width << nlStr;
+  }
+
+  // Generates an input declaration given a sort lid and a name
+  void genInput(size_t width, StringRef name) {
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Generate input declaration
+    os << lid++ << wsStr << inputStr << wsStr << sid << wsStr << name << nlStr;
+  }
+
+  // Generates a constant declaration given a value, a width and a name
+  void genConst(int64_t value, size_t width, Operation *op) {
+    // For now we're going to assume that the name isn't taken, given that hw is
+    // already in SSA form
+    getOrCreateOpLID(op);
+
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    os << lid++ << wsStr << constdStr << wsStr << sid << wsStr << value
+       << nlStr;
+  }
+
+  // Generates a zero constant expression
+  void genZero(size_t width) {
+    // Check if the constant has been created yet
+    if (getConstLID(0, width) != noLID)
+      return;
+
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Keep track of this value in a constant declaration tracker
+    constToLIDMap[{0, width}] = lid;
+
+    // Build and return the zero btor instruction
+    os << lid++ << wsStr << zeroStr << wsStr << sid << nlStr;
+  }
+
+  // Generates a binary operation instruction given an op name, two operands and
+  // a result width
+  void genBinOp(StringRef inst, Operation *binop, Value op1, Value op2,
+                size_t width) {
+    // Set the LID for this operation
+    getOrCreateOpLID(binop);
+
+    // Find the sort's lid
+    size_t sid = sortToLIDMap.at(width);
+
+    // Assuming that the operands were already emitted
+    // Find the LIDs associated to the operands
+    size_t op1LID = getOrCreateOpLID(op1);
+    size_t op2LID = getOrCreateOpLID(op2);
+
+    // Build and return the string
+    os << lid++ << wsStr << inst << wsStr << sid << wsStr << op1LID << wsStr
+       << op2LID << nlStr;
+  }
```

Comment:
I like this pattern of storing the operand LIDs and then using them for printing. Could you use this also for the LID of the op you're currently emitting?
```
auto lid = getOrCreateOpLID(binop); // would also increment the internal lid counter
...
os << lid << " " << inst << ...;
```
Right now the mutation of `lid` in the background isn't very obvious in the code: `getOrCreateOpLID` uses the current `lid` to associate the op, and the code is then implicitly assumed to not mutate `lid` in any way until the `os << ...` line, and for no other calls to `getOrCreateOpLID` to use that same `lid` for an association with `op1` and `op2`. It might be more obvious if `getOrCreateOpLID` was responsible for handing out new LIDs, and if the fact that it uses a counter variable internally could just be an implementation detail. You'd then just ask `getOrCreateOpLID` for the LID of an op, and you either get a pre-existing one or a new one, and then use that.

Since this assumes that `op1` and `op2` have already been emitted and assigned a LID, I think this really needs a `getOpLID` function that screams loudly if that asusmption breaks.

Reply:
The problem with this method is that unfortunately not all btor2 instructions have an associated mlir op, so doing it this way makes the way lids are incremented very inconsistent (sometimes it's on generation, sometimes it's when the op is associated to the lid).

Reply:
I like the `getOrCreate` approach itself, because you're basically saying "give me whatever ID you have for this op, or assign it one if you haven't done so already". That has a nice contract of passing in the op, and then getting an ID, without any possibility for abuse (manually setting the ID, incrementing it inadvertently, etc.). This seems to make sense for the call to `getOrCreate` with the op that's currently being emitted. For operands however, it makes sense to have a `get` flavor which blows up very loudly if the operand does not have a lid assigned. That'll strongly encode the assumption of handling operands before users of the operands.